lightning/src/ln/channelmonitor.rs

```diff
@@ -159,77 +165,174 @@ impl_writeable!(HTLCUpdate, 0, { payment_hash, payment_preimage, source });
 /// index by a PublicKey which is required to sign any updates.
 ///
 /// If you're using this for local monitoring of your own channels, you probably want to use
-/// `OutPoint` as the key, which will give you a ManyChannelMonitor implementation.
-pub struct SimpleManyChannelMonitor<Key, ChanSigner: ChannelKeys, T: Deref, F: Deref, L: Deref, C: Deref>
-	where T::Target: BroadcasterInterface,
+/// `OutPoint` as the key, which will give you a [`chain::Watch`] implementation.
+///
+/// [`chain::Watch`]: ../../chain/trait.Watch.html
+pub struct ChainMonitor<Key, ChanSigner: ChannelKeys, C: Deref, T: Deref, F: Deref, L: Deref>
+	where C::Target: chain::Notify,
+        T::Target: BroadcasterInterface,
         F::Target: FeeEstimator,
         L::Target: Logger,
-        C::Target: ChainWatchInterface,
 {
 	#[cfg(test)] // Used in ChannelManager tests to manipulate channels directly
 	pub monitors: Mutex<HashMap<Key, ChannelMonitor<ChanSigner>>>,
 	#[cfg(not(test))]
 	monitors: Mutex<HashMap<Key, ChannelMonitor<ChanSigner>>>,
-	chain_monitor: C,
+	watch_events: Mutex<WatchEventCache>,
+	chain_source: Option<C>,
 	broadcaster: T,
 	logger: L,
 	fee_estimator: F
 }
 
-impl<Key : Send + cmp::Eq + hash::Hash, ChanSigner: ChannelKeys, T: Deref + Sync + Send, F: Deref + Sync + Send, L: Deref + Sync + Send, C: Deref + Sync + Send>
-	ChainListener for SimpleManyChannelMonitor<Key, ChanSigner, T, F, L, C>
-	where T::Target: BroadcasterInterface,
+struct WatchEventCache {
+	watched: ChainWatchedUtil,
+	events: Vec<WatchEvent>,
+}
+
+/// An event indicating on-chain activity to watch for pertaining to a channel.
```

Comment:
Given we're already calling flush_events reentrant-ly here, does it really make sense to build a bit cache thing? Or is this just to do the filtering inside ChainMonitor (and it will be removed in a followup?)? I'm a bit confused why we can't just drop this now if there's filtering at the callsite in tests.

Reply:
> preventing duplicate calls to chain::Notify for the same watch events
> determining if any new watch events were generated (and thus what block_connected should return)

Hmm, can't the ChannelMonitor remove the duplicate calls itself since it now has the full monitoring data? Then if there were any returned values you know you need to rescan.

> I see your point regarding filtering. But if I remove filtering from block_connected, I see a few test failures:

Right, 99% those are just the tests being overly strict, which should be fixed. Totally fine to do it separately, I'm just dubious we can't avoid this new struct even without addressing that.

Reply:
> Hmm, can't the ChannelMonitor remove the duplicate calls itself since it now has the full monitoring data? Then if there were any returned values you know you need to rescan.

Could you clarify what you mean by "since it now has the full monitoring data"? I didn't change `ChannelMonitor` materially.

Reply:
Right, it's new in that its from Jan of this year in 73dce207dd0ea6c3ac57af3ebb8b87ee03e82c9e.

Reply:
I was able to remove `WatchEventCache` -- and also filtering -- and use the full monitoring data without difficulty, but I'm having trouble with the tests. For `test_bump_penalty_txn_on_revoked_htlcs`, the number of broadcasted transactions is reduced by one with the inputs of `node_txn[2]` and `node_txn[3]` combined into a single transaction (`node_txn[2]`). I can adjust [those assertions](https://github.com/rust-bitcoin/rust-lightning/blob/478c481c7f9c8fc10669fb6b9d2f2c16751a432c/lightning/src/ln/functional_tests.rs#L7877-L7880), but the `check_spends` that follows fails. And commenting it out gives an overflow when subtracting in the `fee_1` calculation if replacing `node_txn[3]` with `node_txn[2]` since the output value is larger now with 3 inputs instead of 2.

Any insight into why the two transactions are combined into one and how to go about modifying this test?

Reply:
Hmm, not immediately, it may be that the filtering in the monitors is broken. Can you push the WIP commit(s) somewhere I can play with?

Reply:
Change is in 71c76068b5e1f368d52e335e48d704a34878b2c1 and the attempt fix in 5acef01c437b88f7f88ede0ae2323892ac863658.

Reply:
Looks like its claiming a third (revoked) output - https://github.com/TheBlueMatt/rust-lightning/commits/2020-08-check-spends-fees calculates the fees more robustly and uses them for this specific bug. I'm not actually sure *why* its now spending the commitment-tx revoked output in addition to the HTLC revoked outputs, though. Maybe @ariard can shed some light?